src/atomWithHash.ts

```diff
@@ -15,16 +15,52 @@ const safeJSONParse = (initialValue: unknown) => (str: string) => {
   }
 };
 
+export type SetHashOption =
+  | 'default'
+  | 'replaceState'
+  | ((searchParams: string) => void);
+
+export type AtomWithHashSetOptions = {
+  setHash?: SetHashOption;
+};
+
+export const setHashWithPush = (searchParams: string) => {
+  window.location.hash = searchParams;
+};
+
+export const setHashWithReplace = (searchParams: string): void => {
+  window.history.replaceState(
+    window.history.state,
+    '',
+    `${window.location.pathname}${window.location.search}#${searchParams}`,
+  );
+};
+
+function getSetHashFn(setHashOption?: SetHashOption) {
+  if (setHashOption === 'replaceState') {
+    return setHashWithReplace;
+  }
+  if (typeof setHashOption === 'function') {
+    return setHashOption;
+  }
+  return setHashWithPush;
+}
+
 export function atomWithHash<Value>(
   key: string,
   initialValue: Value,
   options?: {
     serialize?: (val: Value) => string;
     deserialize?: (str: string) => Value;
     subscribe?: (callback: () => void) => () => void;
-    setHash?: 'default' | 'replaceState' | ((searchParams: string) => void);
+    setHash?: SetHashOption;
   },
-): WritableAtom<Value, [SetStateActionWithReset<Value>], void> {
+): WritableAtom<
+  Value,
+  | [SetStateActionWithReset<Value>]
+  | [SetStateActionWithReset<Value>, AtomWithHashSetOptions],
```

Comment:
This could probably get written as one type like this:
```js
[SetStateActionWithReset<Value>, AtomWithHashSetOptions?]
```
This would better match how atomWithLocation is defined.

Reply:
TIL! the tuple union was my naive first try, didn't know you can do this!

Reply:
pushed this change!